src/daemon/plugin.c

```diff
@@ -1085,48 +1091,78 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+  /* If AlignRead is not set, start reading immediately. */
+  if (alignRead < 0) {
     rf->rf_next_read = cdtime();
-  } else {
-    cdtime_t now = cdtime();
-
-    /* get minute in hour and second in minute and millisecond in second */
-    struct timeval tv = CDTIME_T_TO_TIMEVAL(now);
-    uint64_t curr_msec =
-        (((tv.tv_sec / 60) % 60) * 60 // minute of hour in seconds
-         + gmtime(&(tv.tv_sec))->tm_sec) *
-            1000             // add seconds of minute and convert to ms
-        + tv.tv_usec / 1000; // add millisecond of second
-
-    /* DEBUG("%s: current time: %lu sec, %lu ms in minute, start time: %.3lf,
-    interval: %lu", rf->rf_name, tv.tv_sec, curr_msec, start,
-    CDTIME_T_TO_MS(rf->rf_interval)); */
-
-    /* determine delay until next (aligned) read */
-    int64_t sleep_cdt = DOUBLE_TO_CDTIME_T(start) - MS_TO_CDTIME_T(curr_msec);
-    while (sleep_cdt < 0) {
-      sleep_cdt += rf->rf_interval;
+    return;
+  }
+
+  /* AlignRead values greater than Interval do not make sense. */
+  if (alignRead >= rf->rf_interval) {
+    static bool align_error = true;
+
+    if (align_error) {
+      ERROR("AlignRead has to be smaller than Interval (%.3lf >= %d)!",
+           alignRead, (int)rf->rf_interval);
+      align_error = false;
     }
 
-    /* set the next read time */
-    if (sleep_cdt > 0) {
-      rf->rf_next_read = now + sleep_cdt;
+    rf->rf_next_read = cdtime();
+    return;
+  }
 
-      DEBUG("%s plugin: wait %.3fs before first read at %ld.%03ds since epoche",
-            rf->rf_name, CDTIME_T_TO_DOUBLE(sleep_cdt),
-            (long int)(CDTIME_T_TO_TIMEVAL(rf->rf_next_read).tv_sec),
-            (int)(CDTIME_T_TO_TIMEVAL(rf->rf_next_read).tv_usec / 1000));
-    } else {
-      rf->rf_next_read = cdtime();
+  /* Seconds and minutes of Interval and AlignRead have to be perfect divisors
+   * of 60. Otherwise to perserve alignment. */
+  int sec_interval = CDTIME_T_TO_TIME_T(rf->rf_interval) % 60;
+  int sec_align = (long)alignRead % 60;
+  int min_interval = CDTIME_T_TO_TIME_T(rf->rf_interval) / 60;
+  int min_align = (int)(alignRead / 60);
+  if ((sec_interval > 0 && 60 % sec_interval != 0) ||
+      (sec_align > 0 && 60 % sec_align != 0) ||
+      (min_interval > 0 && 60 % min_interval != 0) ||
+      (min_align > 0 && 60 % min_align != 0)) {
+    static bool align_error = true;
+
+    if (align_error) {
+      ERROR("Seconds and minutes of AlignRead and Interval have to be perfect "
+            "divisors of 60 to perserve alignment.");
+      align_error = true;
     }
+
+    rf->rf_next_read = cdtime();
+    return;
   }
+
+  /* get minutes in current hour and seconds in current minute */
+  cdtime_t now = cdtime();
+  struct timeval now_tv = CDTIME_T_TO_TIMEVAL(now);
+  struct tm *now_tm;
+  now_tm = localtime(&now_tv.tv_sec);
+
+  /* put current minutes, seconds and fractions of a second in a microseconds
+   * value (with stripped hours, days, etc.) */
+  uint64_t now_us =
```

Comment:
Ouch.

All you need here is:

- calculate `now_us` as microseconds since hour start
- compare it with `alignRead`. if `alignRead > now_us`, then next time is:
- now_tv - now_us + alignRead

You must learn how `cdtime_t` works and use it only.

Reply:
That's what the code does, if you remove the `if (sleep_cdt < 0)` block. The if-block considers the case `alignRead < now_us`.

Reply:
`now - us_since_hour_start + align_since_hour_start` is MUCH more understandable than ` sleep_cdt +=  rf->rf_interval * ceil((double)(-sleep_cdt) / (double)rf->rf_interval);` and does not require floating point math. Much understandable also means less error prone.

You need to learn how `cdtime_t` works and use it.

Reply:
The case you are talking about is `alignRead < now_us`, which does not require floating point math in my code. The other case is a little bit more complicated, as we want to find the shortest delay until the next aligned time (as written in the comment). But you are right, I can give `now_us` a more self-explanatory name.

Reply:
```
  /* determine the minimum delay until next (aligned) read */
  int64_t sleep_cdt = DOUBLE_TO_CDTIME_T(alignRead) - US_TO_CDTIME_T(now_us);
  if (sleep_cdt < 0) {
    sleep_cdt +=
        rf->rf_interval * ceil((double)(-sleep_cdt) / (double)rf->rf_interval);
  }

  /* set the next read time */
  rf->rf_next_read = now + sleep_cdt;
```

This piece of code uses floating point.
----
If `sleep_cdt >=0`:  

1) sleep_cdt = alignRead - now_us;
2) rf_next_read = now + sleep_cdt  = now - now_us + alignRead.

where 
`now_us` is usec since hour start 
`now - now_us` is 'startOfHour' point 
and 
`alignRead` is alignRead + interval * 0  (by the fact, for first reading in hour).

---

these expressions lead us to:

The `sleep_cdt < 0` is only when `now_us > alignRead` i.e. first aligned point of hour is in past.
We need to know next aligned point (numbered as X starting at 0). 
Then math is nearby same:

sleep_cdt = now - now_us + alignRead + interval * X 

```
if (now_us > alignRead) {
sleep_cdt = now - now_us + alignRead + interval_us * int(now_us / interval_us) 
}
```
this math not require floating point calculation and that's why is better.

Thinking further, the `if (now_us > alignRead) {}` condition is not correct too.
The correct one is to compare `now_us` with `interval_us`, because `alignRead` could not be >=  `interval_us`.

So we don't need this conditional here at all.

`sleep_cdt = now - now_us + alignRead + interval_us * int(now_us / interval_us)` fully covers all valid cases. Recall what `now - now_us` is `startOfHour`.

If `now_us < interval_us` then `int(now_us / interval_us)` will be 0 and first read will be at `startOfHour + alignRead * 1` which is expected.

If `alignRead == 0`  then all would be correct too.
Pure math. (cd_t and _us variables are mixed to simplify the appearance of the formulas).